sys/kern/src/arch/arm_m.rs

```diff
@@ -349,6 +350,99 @@ pub fn reinitialize(task: &mut task::Task) {
+    // This platform requires 32-byte alignment of all regions.
+    if base & 0x1F != 0 {
+        panic!();
+    }
+
+    let ratts = attributes;
+    let xn = !ratts.contains(RegionAttributes::EXECUTE);
+    // These AP encodings are chosen such that we never deny *privileged*
+    // code (i.e. us) access to the memory.
+    let ap = if ratts.contains(RegionAttributes::WRITE) {
+        0b011
+    } else if ratts.contains(RegionAttributes::READ) {
+        0b010
+    } else {
+        0b001
+    };
+    // Set the TEX/SCB bits to configure memory type, caching policy, and
+    // shareability (with other cores or masters). See table B3-13 in the
+    // ARMv7-M ARM. (Settings are identical on v6-M but the sharability and
+    // TEX bits tend to be ignored.)
+    let (tex, scb) = if ratts.contains(RegionAttributes::DEVICE) {
+        // Device memory.
+        (0b000, 0b001)
+    } else if ratts.contains(RegionAttributes::DMA) {
+        // Conservative settings for normal memory assuming that DMA might
+        // be a problem:
+        // - Outer and inner non-cacheable.
+        // - Shared.
+        (0b001, 0b100)
+    } else {
+        // Aggressive settings for normal memory assume that it is used only
+        // by this processor:
+        // - Outer and inner write-back
+        // - Read and write allocate.
+        // - Not shared.
+        (0b001, 0b011)
+    };
+    // On v6/7-M the MPU expresses size of a region in log2 form _minus
+    // one._ So, the minimum allowed size of 32 bytes is represented as 4,
+    // because `2**(4 + 1) == 32`.
+    //
+    // We store sizes in the region table in an architecture-independent
+    // form (number of bytes) because it simplifies basically everything
+    // else but this routine. Here we must convert between the two -- and
+    // quickly, because this is called on every context switch.
+    //
+    // The image-generation tools check at build time that region sizes are
+    // powers of two. So, we can assume that the size has a single 1 bit. We
+    // can cheaply compute log2 of this by counting trailing zeroes, but
+    // ARMv7-M doesn't have a native instruction for that -- only leading
+    // zeroes. The equivalent using leading zeroes is
+    //
+    //   log2(N) = bits_in_word - 1 - clz(N)
+    //
+    // Because we want log2 _minus one_ we compute it as...
+    //
+    //   log2_m1(N) = bits_in_word - 2 - clz(N)
+    //
+    // If the size is zero or one, this subtraction will underflow. This
+    // should not occur in a valid image, but could occur due to runtime
+    // flash corruption. Any region size under 32 bytes is illegal on
+    // ARMv7-M anyway, so panicking is better than triggering possibly
+    // undefined hardware behavior.
+    //
+    // On ARMv6-M, there is no CLZ instruction either. This winds up
+    // generating decent intrinsic code for `leading_zeros` so we'll live
+    // with it.
+    let l2size = 30 - size.leading_zeros();
```

Comment:
That comment might have been unclear, I'm using `leading_zeros` there _despite_ it not having native support on v6M, because the code generation is pretty okay. You are correct, of course, that there should no longer _be any_ code generation.

`leading_zeros` and `trailing_zeros` here are not interchangeable; `trailing_zeros` would measure the smallest power-of-two alignment of the size, not the magnitude of the size itself.

Reply:
Oops, yeah I got thrown off and started thinking that this was a case of `30 - size.leading_zeros() == size.trailing_zeros()`.